### DISCUSSION TITLE
Update UI from EventDispatcher::schedule?

### DISCUSSION THREAD

**Discussion by qingxiang-jia** [fcitx/fcitx5#916] [Q&A] [answered]
Summary: updating UI from EventDispatcher doesn't seem to work (no UI is shown).

Short version: I created an EventDispatcher, attach it to instance's event loop. I call EventDispatcher::schedule with a lambda from a different thread, where I update the UI. Result: UI is not shown, but the UI update code inside lambda is called.

Longer version: I have two processes (A and [B](https://github.com/qingxiang-jia/full-cloud-pinyin/tree/main/fcp-zmq/fcp)). A is an Fcitx [engine](https://github.com/qingxiang-jia/full-cloud-pinyin/blob/main/fcp-zmq/ims/ims.h), call it ImsEngine, it publishes key event to ZMQ and another process (B) listens. ImsEngine creates [ImsServer](https://github.com/qingxiang-jia/full-cloud-pinyin/blob/main/fcp-zmq/ims/ims.h) and it receives candidates from another process B (via ZMQ). ImsServer has a pointer to the event dispatcher, and runs in another thread. Whenever ImsServer receives candidates from ZMQ, it uses the event [dispatcher](https://github.com/qingxiang-jia/full-cloud-pinyin/blob/main/fcp-zmq/ims/ims.cpp#L339) to schedule an UI update. However, while the UI update code runs, the UI isn't shown.

Thank you for the help!

**Comment by qingxiang-jia** [comment]
I am not sure if this is the right solution, but it does the work:

Use EvenDispatcher::schedule with a nullptr argument to wake up the event loop (the one from Instance when ImsEngine is initialized), and then do the UI update.

**Comment by wengxt** [comment]
What's the version for fcitx5 and what's your environment? Can you provide the output of fcitx5-diagnose

In some older version, there's indeed some bug that the xcb/wayland ui is not flushing the connection so Ui didn't really update.

**Reply by qingxiang-jia** [reply]
I encounter this issue on both master (7d528e0821116c04a6a74f4194bc072864b3f0b9) and fcitx5 5.1.5-1 (Arch).

I included output of fcitx5-diagnose for both master and fcitx5 5.1.5-1. For the master version, I only built fcitx5 and fcitx5-gtk to test the issue (so it might say other components missing).
[diagnosis-5.1.5-1.txt](https://github.com/fcitx/fcitx5/files/13722079/diagnosis-5.1.5-1.txt)
[diagnosis-master.txt](https://github.com/fcitx/fcitx5/files/13722080/diagnosis-master.txt)

**Reply by wengxt** [reply]
I don't think your code is correct.

Actually, you should not touch fcitx data in non-fcitx thread, unless it is explicitly said it's thread-safe. 

This includes input context and any data associated to input context. You should do all UI update in dispatcher's callback.

**Reply by qingxiang-jia** [reply]
You are looking at my updated main branch. The updated version indeed just wake up the event loop and access input context directly from a different thread, so not correct. However, this setup works.

The original [version](https://github.com/qingxiang-jia/full-cloud-pinyin/commit/110e0e904102c0b830b70e407ec6ab675d6db19a), I put all code inside dispatcher's callback and it doesn't work (that's why we had this discussion).

**Reply by qingxiang-jia** [reply]
BTW, I am sure you know about this, but just in case. When putting UI update code in dispatcher's callback, the UI isn't really updated, because [ClassicUI::update](https://github.com/fcitx/fcitx5/blob/master/src/ui/classic/classicui.cpp#L412) is not called. I don't know why it's the case, but hopefully this offers a bit more information.

**Reply by wengxt** [reply]
The key issue is the same, you're still not doing it fully in main thead.

For example, I noticed following backtrace:
```
ent&, fcitx::InputContext*&&, bool&) (this=0x556cdd6f3330) at /home/csslayer/Develop/fcitx5/src/lib/fcitx/inputcontext_p.h:45
#18 0x00007f1bd02e973b in fcitx::InputContext::updateUserInterface(fcitx::UserInterfaceComponent, bool)
    (this=0x556cdd746960, component=fcitx::UserInterfaceComponent::InputPanel, immediate=false) at /home/csslayer/Develop/fcitx5/src/lib/fcitx/inputcontext.cpp:468
#19 0x00007f1ba9686843 in QuweiEngine::uiUpdate() (this=0x556cde6530f0) at /home/csslayer/Develop/full-cloud-pinyin/fcp-fcitx5/fcitx5/src/quwei.cpp:371
#20 0x00007f1ba96864e5 in QuweiEngine::setCandidates(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >) (this=0x556cde6530f0, candidates=std::vector of length 161, capacity 161 = {...})
    at /home/csslayer/Develop/full-cloud-pinyin/fcp-fcitx5/fcitx5/src/quwei.cpp:349
#21 0x00007f1ba9684508 in set_candidates(int16_t**, size_t) (candidates=0x7f1b8c02fdf0, cnt=161) at /home/csslayer/Develop/full-cloud-pinyin/fcp-fcitx5/fcitx5/src/quwei.cpp:48
#22 0x00007f1ba969a966 in fcp::fcitx5::Fcitx5::ui_set_candidates (self=0x556cdd6c88b8, display_texts=...) at src/fcitx5.rs:25
#23 0x00007f1baa2c56bb in fcp::fcp::{impl#2}::on_key_press::{async_block#3} () at src/fcp.rs:447
#24 0x00007f1baa2def0c in core::future::from_generator::{impl#1}::poll<fcp::fcp::{impl#2}::on_key_press::{async_block_env#3}> (self=..., cx=0x7f1b97ffd408)
 ```

This is update not called from main thread. And you're triggering a systemd assertion and crash.

Please check and make sure all the relevant function are using the event dispatcher.

**Reply by qingxiang-jia** [reply]
Good morning and I am sorry for the confusion. You have been looking at a wrong place (but that's my fault, I didn't clearly state which directory you should look into, and the README is outdated, again I am sorry). Yes, what you said were all right, but that's old code (and I should have deleted to avoid the confusion), again, I apologize for letting you waste time on wrong place.

The code is at:
https://github.com/qingxiang-jia/full-cloud-pinyin/tree/main/fcp-zmq/ims (Fcitx5 part)
https://github.com/qingxiang-jia/full-cloud-pinyin/tree/main/fcp-zmq/fcp (Rust part)

To compile ims, you need ZMQ (if you are using Arch, the one in repository is good). I have a Makefile to simplify dev process. You can use:
`make init` to initialize cmake.
`make build` to build the code.
`make install` to install it to /usr/local (I have a debug build of fcitx5 and fctix5-gtk also in /usr/local)

For the Rust side, just run `cargo run`. For now, you need to run the Rust side after fcitx5 is running.

Also, I made a branch for you to check: https://github.com/qingxiang-jia/full-cloud-pinyin/tree/for-fcitx5-author-thank-you

**Reply by wengxt** [reply, marked as answer]
This branch has a very obvious bug, the cmd is a variable on stack and schedule lambda capture the pointer to it.

**Reply by qingxiang-jia** [reply]
Oh shoot, you are right 😭😭😭.
